src/scheduler/MpiWorld.cpp

```diff
@@ -912,53 +918,138 @@ void MpiWorld::gather(int sendRank,
 
     bool isInPlace = sendBuffer == recvBuffer;
 
-    // If we're the root, do the gathering
```

Comment:
The code is now lacking comments like this that say what it's doing at a high level. At the start of the function you could put a comment summarising the four (?) different scenarios, then make it clearer by factoring out some of the booleans to make the big conditional easier to parse:

```cpp
bool isGatherRoot = sendRank == recvRank;
bool isLocalLeader = sendRank == localLeader;
bool rootOnThisHost = getHostForRank(recvRank) == thisHost;

if(isGatherRoot) {
    // If this is the root rank of the gather, do xyz
    ...
} else if(isLocalLeader && rootOnThisHost) {
    // If this_ rank is local leader and the root rank is on this host too, we do xyz
} else if(isLocalLeader) {
    // If this rank is the local leader and root is on another host we do xyz
} else {
    // If this rank is not local leader or root, do xyz
}
```

Reply:
Yes, I actually think this was a very good idea and it makes the code much more readable (imo).